File: leetcode/101-200/109-convert-sorted-list-to-binary-search-tree/bench/sorted_list_to_bst.c
// Benchmark workload for LeetCode #109 — sorted list to BST, C mirror (malloc *Node).
// Build 8 sorted linked lists once (kept alive across reps), then K reps of the array-
// conversion sorted_list_to_bst on a data-dependent-selected list: walk the list into a local
// array, build a fresh balanced tree by middle-pick, fold its serialization, free the tree.
#include <stdio.h>
#include <stdlib.h>
#define MOD 1000000007LL

typedef struct LNode { long long val; struct LNode* next; } LNode;
typedef struct TNode { long long val; struct TNode *left, *right; } TNode;

static LNode* build_list(long long len, long long off) {
    LNode* head = NULL;
    for (long long i = len - 1; i >= 0; i--) {
        LNode* n = malloc(sizeof(LNode));
        n->val = off + 1 + i;
        n->next = head;
        head = n;
    }
    return head;
}
static TNode* build_from_arr(const long long* arr, long long lo, long long hi) {
    if (lo > hi) return NULL;
    long long mid = (lo + hi) / 2;
    TNode* n = malloc(sizeof(TNode));
    n->val = arr[mid];
    n->left  = build_from_arr(arr, lo, mid - 1);
    n->right = build_from_arr(arr, mid + 1, hi);
    return n;
}
static TNode* sorted_list_to_bst(LNode* head) {
    long long arr[64];
    long long n = 0;
    for (LNode* c = head; c; c = c->next) arr[n++] = c->val;
    return build_from_arr(arr, 0, n - 1);
}
static long long ser(TNode* n, long long acc) {
    if (!n) return (acc * 131 + 1) % MOD;
    acc = (acc * 131 + (n->val + 2)) % MOD;
    acc = ser(n->left, acc);
    acc = ser(n->right, acc);
    return acc;
}
static void free_tree(TNode* n) { if (!n) return; free_tree(n->left); free_tree(n->right); free(n); }
static void free_list(LNode* h) { while (h) { LNode* nx = h->next; free(h); h = nx; } }

int main(void) {
    LNode* pool[8];
    for (long long t = 0; t < 8; t++) pool[t] = build_list(15, t * 100);
    long long acc = 1;
    for (long long rep = 0; rep < 1000000; rep++) {
        long long idx = acc % 8;
        TNode* root = sorted_list_to_bst(pool[idx]);
        long long s = ser(root, 0);
        free_tree(root);
        acc = (acc * 131 + s) % MOD;
    }
    printf("%lld\n", acc);
    for (long long t = 0; t < 8; t++) free_list(pool[t]);
    return 0;
}

File: leetcode/101-200/109-convert-sorted-list-to-binary-search-tree/bench/sorted_list_to_bst.rs
// Benchmark workload for LeetCode #109 — sorted list to BST, Rust mirror.
// The input list is a shared, immutable structure read many times, so `Rc<ListNode>` is the
// honest Rust list (reference-counted, kept alive across reps like Kāra's `shared`); the output
// tree is single-owner `Box` (built then dropped). Build 8 lists once, then K reps of the
// array-conversion sorted_list_to_bst on a data-dependent-selected list.
use std::rc::Rc;
const MOD: i64 = 1000000007;

struct LNode { val: i64, next: Option<Rc<LNode>> }
struct TNode { val: i64, left: Option<Box<TNode>>, right: Option<Box<TNode>> }

fn build_list(len: i64, off: i64) -> Option<Rc<LNode>> {
    let mut head: Option<Rc<LNode>> = None;
    for i in (0..len).rev() {
        head = Some(Rc::new(LNode { val: off + 1 + i, next: head.take() }));
    }
    head
}
fn build_from_arr(arr: &[i64], lo: i64, hi: i64) -> Option<Box<TNode>> {
    if lo > hi { return None; }
    let mid = (lo + hi) / 2;
    Some(Box::new(TNode {
        val: arr[mid as usize],
        left: build_from_arr(arr, lo, mid - 1),
        right: build_from_arr(arr, mid + 1, hi),
    }))
}
fn sorted_list_to_bst(head: &Option<Rc<LNode>>) -> Option<Box<TNode>> {
    let mut arr: Vec<i64> = Vec::new();
    let mut cur = head.clone();
    while let Some(n) = cur {
        arr.push(n.val);
        cur = n.next.clone();
    }
    build_from_arr(&arr, 0, arr.len() as i64 - 1)
}
fn ser(node: &Option<Box<TNode>>, acc: i64) -> i64 {
    match node {
        None => (acc * 131 + 1) % MOD,
        Some(n) => {
            let mut a = (acc * 131 + (n.val + 2)) % MOD;
            a = ser(&n.left, a);
            a = ser(&n.right, a);
            a
        }
    }
}

fn main() {
    let mut pool: Vec<Option<Rc<LNode>>> = Vec::new();
    for t in 0..8i64 { pool.push(build_list(15, t * 100)); }
    let mut acc: i64 = 1;
    for _ in 0..1000000i64 {
        let idx = (acc % 8) as usize;
        let root = sorted_list_to_bst(&pool[idx]);
        let s = ser(&root, 0);
        acc = (acc * 131 + s) % MOD;
    }
    println!("{}", acc);
}
